src/murmur/Cert.cpp

```diff
@@ -15,12 +15,109 @@
 
 #include <openssl/err.h>
 #include <openssl/evp.h>
+#include <openssl/ssl.h>
 #include <openssl/x509.h>
 
 #ifdef Q_OS_WIN
 #	include <winsock2.h>
 #endif
 
+QList< QSslCertificate > Server::buildSslChain(const QSslCertificate &leaf, const QList< QSslCertificate > &pool) {
+	QList< QSslCertificate > chain;
+	if (leaf.isNull()) {
+		return chain;
+	}
+	chain << leaf;
+	if (pool.isEmpty()) {
+		return chain;
+	}
+
+	// Convert the leaf to der format and create an openssl x509 from it
+	QByteArray qbaLeaf = leaf.toDer();
+	int maxDerSize     = qbaLeaf.size();
+	BIO *mem           = BIO_new_mem_buf(qbaLeaf.data(), maxDerSize);
+	Q_UNUSED(BIO_set_close(mem, BIO_NOCLOSE));
+	X509 *leaf_x509 = d2i_X509_bio(mem, nullptr);
+	BIO_free(mem);
+
+	// Prepare a ssl context, the method should not matter, so just go with TLS_method()
+	SSL_CTX *ctx = SSL_CTX_new(TLS_method());
+
+	// Add the leaf
+	SSL_CTX_use_certificate(ctx, leaf_x509);
+
+	// Construct openssl x509 for the pool and add each to the ctx
+	foreach (const QSslCertificate &cert, pool) {
+		QByteArray qbaCert = cert.toDer();
+		int s              = qbaCert.size();
+		maxDerSize         = maxDerSize < s ? s : maxDerSize;
+		BIO *mem           = BIO_new_mem_buf(qbaCert.data(), s);
+		Q_UNUSED(BIO_set_close(mem, BIO_NOCLOSE));
+		X509 *x509 = d2i_X509_bio(mem, nullptr);
+		BIO_free(mem);
+		SSL_CTX_add0_chain_cert(ctx, x509);
+	}
+
+	// Do the actual chain building
+	int flags = SSL_BUILD_CHAIN_FLAG_CHECK | SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR; // Think of the correct flags
+	int ret   = SSL_CTX_build_cert_chain(ctx, flags);
+
+	// Check if we were succesfull, since we use the SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR flag
+	// 2 is a acceptable return value, too
+	if (ret == 1 || ret == 2) {
+		// Retrieve the chain
+		STACK_OF(X509) *stack = nullptr;
+		SSL_CTX_get0_chain_certs(ctx, &stack);
+
+		// Copy the chain back to qt
+		// Instead of allocating a new buffer every time i2d_X509 is called, we allocate a shared buffer
+		// and because we know the maxDerSize, we konw how big this needs to be
+		unsigned char *buffer = (unsigned char *) malloc(maxDerSize);
```

Comment:
I would use an `std::vector` instead. It's more elegant and there's no need to free the allocated memory manually.

```suggestion
		std::vector< unsigned char > buffer(maxDerSize);
```

The pointer to the raw buffer would then be `&buffer[0]`.

Reply:
I've never worked with vectors before. Is it safe to pass a pointer to the raw buffer `&buffer[0]` to openssl (as in line 78) (I'm unsure since openssl will try to adjust the pointer)? Will line 86 work the same way? Could i just say:
```c++
std::vector< unsigned char > bufferVector(maxDerSize);
unsigned char *buffer = &bufferVector[0];
```
(and remove the `free` in line 99) to guarantee that the rest is working?

Reply:
Yes. A vector guarantees to store its contents in a contiguous block of memory in order to be usable as a C-style array of the same length.

However in case you don't need the ability to dynamically change the vector's size, I would recommend using a smart-pointer instead of a vector like this:
```cpp
std::unique_ptr< unsigned char[] > buffer = std::make_unique< unsigned char[] >(maxDerSize);
```
This makes it clear that you intend to use a fixed-size array but also makes use of RAII in order to not have to manually call `free` or `delete[]`.

Reply:
So I was thinking of the smart pointer approach. The function `i2d_X509` needs as argument a pointer to a buffer. And a buffer is itself a pointer to the first element of the respective array. `i2d_X509` will then use the pointer to the pointer to the first element it gets, to shift the pointer to the first element (this is while the correction `buffer -= actualSize;` is used below).

As I understand the smart pointer, it is forbidden to access the raw pointer the smart pointer "hides". If we access it in a hacky way, and put a pointer to it `i2d_X509`, we will undermine the concept of a smart pointer, since `i2d_X509` will change the address of the raw pointer in the smart pointer itself. If now an exception occurs before a manual backshift of the raw pointer hidden in the smart pointer to the originall memory region, the smart pointer will try to free the region, but miss the first `actualSize` of memory?

Since we are interacting with openssl here, which is a c library, I'm not sure which c++ concepts we should try to apply here, or if we just stick to the c-way of doing it with malloc and free...